### PR TITLE
Utility functions for NDVarArrays

### PR DESCRIPTION
**Pull request by IgnaceBleukx**
There was an issue with our overriding of numpy-functionality for NDVarArrays with higher dimensions.
E.g., the following happened:
```python
bvs = cp.boolvar(shape=(3,3,2))
res = bvs.any(axis=2)
```
Now `res` was a single clause with 18 elements, while doing the same with a numpy array results a 3x3 matrix, which is correct. This PR fixes that.

[Feature] Also added the option to create common global constraints directly from the array.
E.g., a user can now do `puzzle.alldifferent(axis=1)` to post the alldifferent constraints over the columns of a Sudoku.

### CODE REVIEW

File: cpmpy/expressions/variables.py [hunk 1]
```diff
@@ -468,131 +468,125 @@ def sum(self, axis=None, out=None):
...
-        if axis is None:    # simple case where we want the Minimum over the whole array
-            arr = self.flatten()
-            return Minimum(arr)
-
-        arr = self.__axis(axis=axis)
-
-        out = []
-        for i in range(0, arr.shape[0]):
-            out.append(Minimum(arr[i, ...]))
+        if axis is None:    # simple case where we want the maximum over the whole array
```
Comment (Wout4) [inline cpmpy/expressions/variables.py:517]: minimum

File: cpmpy/expressions/variables.py [hunk 2]
```diff
@@ -468,131 +468,125 @@ def sum(self, axis=None, out=None):
...
-        # return the NDVarArray that contains the Minimum global constraints
-        return out
+        return cpm_array(np.apply_along_axis(cpm_min, axis=axis, arr=self))
 
     def any(self, axis=None, out=None):
         """
             overwrite np.any(NDVarArray)
         """
-        from .python_builtins import any
+        from .python_builtins import max as cpm_any
```
Comment (Wout4) [inline cpmpy/expressions/variables.py:526]: wrong import

File: cpmpy/expressions/variables.py [hunk 3]
```diff
@@ -468,131 +468,125 @@ def sum(self, axis=None, out=None):
...
 
         if out is not None:
             raise NotImplementedError()
 
-        if axis is None:    # simple case where we want the .any() over the whole array
-            arr = self.flatten()
-            return any(arr)
-
-        arr = self.__axis(axis=axis)
+        if axis is None:    # simple case where we want the maximum over the whole array
```
Comment (Wout4) [inline cpmpy/expressions/variables.py:534]: update comment

File: cpmpy/expressions/variables.py [hunk 4]
```diff
@@ -468,131 +468,125 @@ def sum(self, axis=None, out=None):
...
-        # return the NDVarArray that contains the any() constraints
-        return out
 
     def all(self, axis=None, out=None):
         """
             overwrite np.any(NDVarArray)
         """
 
-        from .python_builtins import all
+        from .python_builtins import max as cpm_all
```
Comment (Wout4) [inline cpmpy/expressions/variables.py:545]: wrong import

File: cpmpy/expressions/variables.py [hunk 5]
```diff
@@ -468,131 +468,125 @@ def sum(self, axis=None, out=None):
...
+        if any(not is_boolexpr(x) for x in self.flat):
+            raise TypeError("Cannot call .any() in an array not consisting only of bools")
+
         if out is not None:
             raise NotImplementedError()
 
-        if axis is None:    # simple case where we want the .all() over the whole array
-            arr = self.flatten()
-            return all(arr)
+        if axis is None:  # simple case where we want the maximum over the whole array
```
Comment (Wout4) [inline cpmpy/expressions/variables.py:553]: .all() not maximum

**Review by Wout4 — COMMENTED** [review]
some copy-paste errors, other than that i would maybe add a test case with solve call to make sure it behaves as expected. Same for the global with axis, looks nice though!

**Comment by IgnaceBleukx** [comment]
Fixed the copy-paste errors

File: cpmpy/expressions/variables.py [hunk 6]
```diff
@@ -468,131 +468,125 @@ def sum(self, axis=None, out=None):
...
+
+    def alldifferent(self, axis=None):
+        from .globalconstraints import AllDifferent
+        return cpm_array(np.apply_along_axis(lambda arr : AllDifferent(arr), axis=axis, arr=self))
+
+    def alldifferent_except0(self, axis=None):
+        from .globalconstraints import AllDifferentExcept0
+        return cpm_array(np.apply_along_axis(lambda arr: AllDifferentExcept0(arr), axis=axis, arr=self))
+
+    def allequal(self, axis=None):
```
Comment (Wout4) [inline cpmpy/expressions/variables.py:566]: All of these crash when axis == None (which is the default value).

I guess they could all be valid without specifying an axis right. (would be the same as just calling the global directly)

**Comment by tias** [comment]
I find it quite far-reaching to put all global constraints also in the variables... I've been avoiding this mixing of functionalities and pushing things into the expression objects (e.g. transforms could also be pushed in, which I think is harder to read/maintain, at least I had a lot of diffculty understanding how NumberJack worked in that way)

For numpy we have good reasons: we want to be numpy compatible, so people can treat our arrays as numpy arrays and call numpy functions on them and we want them to produce expressions (do we also call max and min?). But adding our global constraints is not for supporting standard numpy, its "numpy-like" programming of cpmpy.

so, how about only doing the numpy ones?

**Comment by IgnaceBleukx** [comment]
Yes, adding all of the globals was a bit far fetched indeed... I'll adapt it to only include those overriding numpy's builtins

**Review by Wout4 — APPROVED** [review]
This is ready to merge now, axis argument works on all, any, max, min, sum and prod

**Pull request merged**